settings.ts

```diff
@@ -0,0 +1,32 @@
+/** Constants */
+export const YOUTUBE_BASE_URL = "youtube.com";
+export const DEFAULT_TEMPLATE = "---\ndate: {{Date}}\n---\n# {{Title}}\n![]({{ImageURL}})\n## Description:\n{{Description}}\n-> [Youtube video Link]({{PodcastURL}})\n\n## Summary:\n";
+export const DEFAULT_DATE_FORMAT = "YYYY-MM-DD";
+export const DEFAULT_SUMMARY_SIZE = 3;
+export const MAX_SUMMARY_SIZE = 20;
+export const MAX_TOKEN_SIZE_IN_A_REQUEST = 2500;
+
+export interface PluginSettings {
```

Comment:
Also, you don't have to do it as it is out of the scope of this PR, but it will be nice to have a setting for Chat GPT models as requested [here](https://github.com/ozdemir08/youtube-video-summarizer/issues/1). A key difference is that 3.5 models have ~4096 token limits whereas GPT 4 has a much higher limit 128K and we would need to keep it in mind as well.

Feel free to take a look at it as well if you feel like a hero :)

Reply:
Hi @ozdemir08 few questions since I'm not very familiar with how any of the models work or are called:

- Would using different models still use the same OpenAiClient, or would we need support for multiple clients?
- I see `model: 'gpt-3.5-turbo',` specified in `OpenAiClient.ts` -- I'm assuming the settings panel would adjust this value here?
- Went to the [OpenAI docs](https://platform.openai.com/docs/api-reference/models/list) and found this command to fetch a list of all the models.  Do you want to support all of them?  How to determine maxTokenSize for each?

```bash
curl https://api.openai.com/v1/models \
  -H "Authorization: Bearer $OPENAI_API_KEY"
```

```json
{
  "object": "list",
  "data": [
    {
      "id": "gpt-4-turbo-preview",
      "object": "model",
      "created": 1706037777,
      "owned_by": "system"
    },
    {
      "id": "gpt-4-0125-preview",
      "object": "model",
      "created": 1706037612,
      "owned_by": "system"
    },
    {
      "id": "gpt-3.5-turbo-1106",
      "object": "model",
      "created": 1698959748,
      "owned_by": "system"
    },
    {
      "id": "dall-e-2",
      "object": "model",
      "created": 1698798177,
      "owned_by": "system"
    },
    {
      "id": "gpt-3.5-turbo-0125",
      "object": "model",
      "created": 1706048358,
      "owned_by": "system"
    },
    {
      "id": "whisper-1",
      "object": "model",
      "created": 1677532384,
      "owned_by": "openai-internal"
    },
    {
      "id": "tts-1-hd-1106",
      "object": "model",
      "created": 1699053533,
      "owned_by": "system"
    },
    {
      "id": "gpt-3.5-turbo",
      "object": "model",
      "created": 1677610602,
      "owned_by": "openai"
    },
    {
      "id": "tts-1-hd",
      "object": "model",
      "created": 1699046015,
      "owned_by": "system"
    },
    {
      "id": "gpt-4-vision-preview",
      "object": "model",
      "created": 1698894917,
      "owned_by": "system"
    },
    {
      "id": "text-embedding-3-large",
      "object": "model",
      "created": 1705953180,
      "owned_by": "system"
    },
    {
      "id": "gpt-3.5-turbo-0301",
      "object": "model",
      "created": 1677649963,
      "owned_by": "openai"
    },
    {
      "id": "gpt-3.5-turbo-16k",
      "object": "model",
      "created": 1683758102,
      "owned_by": "openai-internal"
    },
    {
      "id": "gpt-4-1106-vision-preview",
      "object": "model",
      "created": 1711473033,
      "owned_by": "system"
    },
    {
      "id": "gpt-4-1106-preview",
      "object": "model",
      "created": 1698957206,
      "owned_by": "system"
    },
    {
      "id": "gpt-3.5-turbo-instruct",
      "object": "model",
      "created": 1692901427,
      "owned_by": "system"
    },
    {
      "id": "gpt-3.5-turbo-instruct-0914",
      "object": "model",
      "created": 1694122472,
      "owned_by": "system"
    },
    {
      "id": "tts-1",
      "object": "model",
      "created": 1681940951,
      "owned_by": "openai-internal"
    },
    {
      "id": "davinci-002",
      "object": "model",
      "created": 1692634301,
      "owned_by": "system"
    },
    {
      "id": "babbage-002",
      "object": "model",
      "created": 1692634615,
      "owned_by": "system"
    },
    {
      "id": "gpt-4-0613",
      "object": "model",
      "created": 1686588896,
      "owned_by": "openai"
    },
    {
      "id": "tts-1-1106",
      "object": "model",
      "created": 1699053241,
      "owned_by": "system"
    },
    {
      "id": "gpt-4",
      "object": "model",
      "created": 1687882411,
      "owned_by": "openai"
    },
    {
      "id": "gpt-3.5-turbo-0613",
      "object": "model",
      "created": 1686587434,
      "owned_by": "openai"
    },
    {
      "id": "text-embedding-ada-002",
      "object": "model",
      "created": 1671217299,
      "owned_by": "openai-internal"
    },
    {
      "id": "gpt-3.5-turbo-16k-0613",
      "object": "model",
      "created": 1685474247,
      "owned_by": "openai"
    },
    {
      "id": "dall-e-3",
      "object": "model",
      "created": 1698785189,
      "owned_by": "system"
    },
    {
      "id": "text-embedding-3-small",
      "object": "model",
      "created": 1705948997,
      "owned_by": "system"
    }
  ]
}
```

- I'm thinking to do these things:
  - in 'settings.ts' export an array of models to choose from
  - in `settings.ts` add an object in the interface for the model named `openAiModel`
  - the object would look something like `{"name": "gpt-3.5-turbo", "maxTokenSize": 2500}`
  - in `settings.ts` export `gpt-3.5-turbo` as the default model
  - the `OpenAIClient.ts` would use `this.settings.openAiModel.name` to specify the name
  - the `TranscriptSummarizer.ts` would use `this.settings.openAiModel.maxTokenSize` to specify the size instead of using `MAX_TOKEN_SIZE_IN_A_REQUEST`
  - in `main.ts` add a new setting dropdown to choose the model from the exported array, which would then set the object with the name & maxTokenSize
  - in `main.ts` update the reset button to set model back to `gpt-3.5-turbo`

Reply:
Hi @ozdemir08, added changes to support model switching in https://github.com/ozdemir08/youtube-video-summarizer/pull/9/commits/77fa3d3485555ed8d08141e3fbf5f2e8da6f37f9

Main difference from what I wrote out above is the data structure for storing the token size.  I went with a map from modelName to tokenSize instead of an array of model objects:

this:
```json
{
	"gpt-3.5-turbo": 4096,
	"gpt-4": 8192,
	"gpt-4-0125-preview": 128000,
}
```

vs. this:

```json
[
  {"name": "gpt-3.5-turbo", "maxTokenSize": 4096},
  {"name": "gpt-4", "maxTokenSize": 8192},
  {"name": "gpt-4-0125-preview", "maxTokenSize": 128000},
]
```

This is my first time being hands-on with TypeScript, so I'm not sure if this makes a big difference or not -- so this would be the area I'd ask you to focus on while reviewing.

Aside from that, it seems to function properly:

![image](https://github.com/ozdemir08/youtube-video-summarizer/assets/750874/ff5da677-0dcb-4d52-b2f6-9150dbe75948)

![image](https://github.com/ozdemir08/youtube-video-summarizer/assets/750874/3075f627-a87a-45de-89d4-b24a9e0a99e9)

![image](https://github.com/ozdemir08/youtube-video-summarizer/assets/750874/27f23bb1-b257-4557-a775-f3f9d4c1ca5b)

![image](https://github.com/ozdemir08/youtube-video-summarizer/assets/750874/e8c79f6f-d491-4247-ae95-1d04371763bf)